src/solvespace.h

```diff
@@ -485,6 +485,48 @@ class GCodeFileWriter : public VectorFileWriter {
     bool CanOutputMesh() const override { return false; }
 };
 
+class Sketch;
+
+class OrderedGroupsIterator {
+public:
+    OrderedGroupsIterator() = default;
+    OrderedGroupsIterator(Sketch &sketch, hGroup *currentGroup)
+        : sk(&sketch), current(currentGroup) {}
+
+    bool IsEnd() const { return current == nullptr; }
+
+    Group *operator*() const;
+
+    // Pre-increment operator
+    // Definition below that for Sketch
+    OrderedGroupsIterator &operator++();
+
+    bool operator==(OrderedGroupsIterator const &other) const {
+        if(IsEnd() && other.IsEnd()) {
+            return true;
+        }
+        return sk == other.sk && current == other.current;
+    }
+    bool operator!=(OrderedGroupsIterator const &other) const { return !((*this) == other); }
+
+private:
+    Sketch *sk      = nullptr;
+    hGroup *current = nullptr;
+};
+
+
+class OrderedGroupsProxy {
```

Comment:
Is there any way to simplify this? Returning a pair or something. As it is I feel like this patch adds more complexity than it removes.

Reply:
I mean, +70 lines -27 lines.

Reply:
I wish there was - but ranges haven't made it into the standard yet. (Maybe c++20?)

No hard feelings, I can close this. Just had it sitting around here.